File: src/lib/theme.ts
import type { User } from "@prisma/client";

// Pulled from the original WPF app's App.xaml SolidColorBrush resources.
export const defaultTheme = {
  background: "#ffeca7",
  accent: "#e49364",
  foreground: "#db434c",
  accentForeground: "#c70039",
} as const;

export interface Palette {
  background: string;
  accent: string;
  foreground: string;
  accentForeground: string;
}

const hexColor = /^#[0-9a-fA-F]{6}$/;

export function isValidHexColor(value: string): boolean {
  return hexColor.test(value);
}

type ThemeFields = Pick<User, "themeBackground" | "themeAccent" | "themeForeground" | "themeAccentForeground">;

/** Falls back to the app default per-field, so a user can override just one color. */
export function paletteForUser(user: ThemeFields | null | undefined): Palette {
  return {
    background: user?.themeBackground ?? defaultTheme.background,
    accent: user?.themeAccent ?? defaultTheme.accent,
    foreground: user?.themeForeground ?? defaultTheme.foreground,
    accentForeground: user?.themeAccentForeground ?? defaultTheme.accentForeground,
  };
}

File: Sources/App/Services/Theme.swift
/// Default palette, taken directly from the original WPF app's `App.xaml`
/// resource dictionary (`MainBackground`, `AccentBackground`,
/// `MainForeground`, `AccentForeground`, `AccentTertiary`).
enum Theme {
    static let defaultBackground = "#ffeca7"
    static let defaultAccent = "#e49364"
    static let defaultForeground = "#db434c"
    static let defaultAccentForeground = "#c70039"

    struct Palette: Encodable {
        let background: String
        let accent: String
        let foreground: String
        let accentForeground: String
    }

    static func palette(for user: User?) -> Palette {
        Palette(
            background: user?.themeBackground ?? defaultBackground,
            accent: user?.themeAccent ?? defaultAccent,
            foreground: user?.themeForeground ?? defaultForeground,
            accentForeground: user?.themeAccentForeground ?? defaultAccentForeground
        )
    }
}
